File: packages/core/crypto.mjs
// Ripple end-to-end crypto — the shared secret is the pairing code, nothing else.
//
// Design goals:
//   * The relay must stay a BLIND pipe. It never sees the raw code and never
//     sees plaintext. So clients route on a HASH of the code and send only
//     sealed (encrypted) blobs.
//   * Every peer that knows the code derives the same key and the same room id,
//     with no key exchange and no accounts.
//
// This module is plain ESM and uses only the Web Crypto API (`crypto.subtle`),
// which behaves identically in modern browsers and in Node >= 20. That's what
// lets us unit-test it here and run the exact same code in the web app.
//
// Threat model honesty: a pairing code has limited entropy, so security rests
// on three things working together — (1) the relay can't see the code (room is
// hashed), (2) deriving the key is deliberately slow (PBKDF2, many iterations),
// which throttles offline guessing, and (3) codes should be strong (long /
// QR-carried). Short human codes remain the weak link; see docs/security.md.

const subtle = globalThis.crypto.subtle;
const enc = new TextEncoder();
const dec = new TextDecoder();

// Fixed, app-wide KDF salt. A per-code salt would add nothing here: the code is
// the only secret and both sides must derive the same key with no exchange. The
// real cost to an attacker is the iteration count below, not the salt.
const KDF_SALT = enc.encode('yap.kdf.v1');
const KDF_ITERATIONS = 210_000;
const IV_BYTES = 12; // AES-GCM standard nonce length

// --- base64url, portable across browser and Node --------------------------
function bytesToB64url(bytes) {
  let bin = '';
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}
function b64urlToBytes(str) {
  const b64 = str.replace(/-/g, '+').replace(/_/g, '/');
  const bin = atob(b64);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

// Normalize a code the same way everywhere, so "k7qf9p" and "K7QF9P" agree.
export function normalizeCode(code) {
  return String(code || '').toUpperCase().replace(/[^A-Z0-9]/g, '');
}

// The relay routes on this — a SHA-256 of the code, never the code itself.
export async function roomFromCode(code) {
  const digest = await subtle.digest('SHA-256', enc.encode(normalizeCode(code)));
  return bytesToB64url(new Uint8Array(digest));
}

// Derive the AES-GCM key from the code. Returns a CryptoKey (non-extractable).
export async function keyFromCode(code) {
  const material = await subtle.importKey(
    'raw', enc.encode(normalizeCode(code)), 'PBKDF2', false, ['deriveKey'],
  );
  return subtle.deriveKey(
    { name: 'PBKDF2', salt: KDF_SALT, iterations: KDF_ITERATIONS, hash: 'SHA-256' },
    material,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt', 'decrypt'],
  );
}

// Encrypt plaintext → a compact base64url blob of (iv || ciphertext).
export async function seal(key, plaintext) {
  const iv = crypto.getRandomValues(new Uint8Array(IV_BYTES));
  const ct = new Uint8Array(
    await subtle.encrypt({ name: 'AES-GCM', iv }, key, enc.encode(plaintext)),
  );
  const out = new Uint8Array(iv.length + ct.length);
  out.set(iv, 0);
  out.set(ct, iv.length);
  return bytesToB64url(out);
}

// Decrypt a blob from seal(). Returns the plaintext, or null if the key is wrong
// or the data was tampered with (AES-GCM authentication fails) — never throws.
export async function unseal(key, blob) {
  try {
    const bytes = b64urlToBytes(blob);
    const iv = bytes.subarray(0, IV_BYTES);
    const ct = bytes.subarray(IV_BYTES);
    const pt = await subtle.decrypt({ name: 'AES-GCM', iv }, key, ct);
    return dec.decode(pt);
  } catch {
    return null;
  }
}

File: packages/core-kt/src/main/kotlin/com/ripple/core/Crypto.kt
package com.ripple.core

import java.nio.charset.StandardCharsets.UTF_8
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Ripple end-to-end crypto — the Kotlin mirror of `packages/core/crypto.mjs` (JS)
 * and `packages/core-rs` (Rust). Byte-for-byte compatible: same room hash, same
 * sealed blob format, so an Android device interoperates with web + desktop with
 * no key exchange. Verified against the shared cross-language vectors (see
 * `CryptoTest` and HANDOFF §3).
 *
 * Pure JVM crypto (`javax.crypto` / `java.security`) — no Android-only APIs — so
 * it runs identically in unit tests and on-device.
 *
 * The shared secret is ONLY the pairing code: both the room id and the AES key are
 * derived from it, so the relay (which sees neither) stays a blind pipe.
 */
object RippleCrypto {
    // Fixed, app-wide KDF salt. DO NOT change — it is part of key derivation and
    // changing it breaks interop with every existing client and the test vectors.
    private val KDF_SALT = "yap.kdf.v1".toByteArray(UTF_8)
    private const val KDF_ITERATIONS = 210_000
    private const val IV_BYTES = 12   // AES-GCM standard nonce length
    private const val TAG_BITS = 128  // AES-GCM auth tag

    private val rng = SecureRandom()
    private val b64urlEnc = Base64.getUrlEncoder().withoutPadding()
    private val b64urlDec = Base64.getUrlDecoder()

    /** Upper-case and strip non-alphanumerics, so "k7qf9p" and "K7QF9P" agree. */
    fun normalizeCode(code: String): String =
        code.uppercase().replace(Regex("[^A-Z0-9]"), "")

    /** base64url(SHA-256(code)). The relay routes on this, never the raw code. */
    fun roomFromCode(code: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(normalizeCode(code).toByteArray(UTF_8))
        return b64urlEnc.encodeToString(digest)
    }

    /** Derive the AES-256-GCM key from the code (PBKDF2-HMAC-SHA256, 210k iters). */
    fun keyFromCode(code: String): SecretKey {
        val spec = PBEKeySpec(normalizeCode(code).toCharArray(), KDF_SALT, KDF_ITERATIONS, 256)
        val bytes = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
            .generateSecret(spec).encoded
        return SecretKeySpec(bytes, "AES")
    }

    /** Encrypt plaintext → compact base64url blob of (iv(12) || ciphertext||tag). */
    fun seal(key: SecretKey, plaintext: String): String =
        sealRaw(key, plaintext, ByteArray(IV_BYTES).also(rng::nextBytes))

    /** Decrypt a [seal] blob; returns null on wrong key / tamper (never throws). */
    fun unseal(key: SecretKey, blob: String): String? = try {
        val bytes = b64urlDec.decode(blob)
        val iv = bytes.copyOfRange(0, IV_BYTES)
        val ct = bytes.copyOfRange(IV_BYTES, bytes.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(TAG_BITS, iv))
        String(cipher.doFinal(ct), UTF_8)
    } catch (e: Exception) {
        null
    }

    /** Seal with a caller-supplied IV — internal, for deterministic vector tests. */
    internal fun sealRaw(key: SecretKey, plaintext: String, iv: ByteArray): String {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(TAG_BITS, iv))
        val ct = cipher.doFinal(plaintext.toByteArray(UTF_8))
        return b64urlEnc.encodeToString(iv + ct)
    }
}
